ARROW-15604: [C++][CI] Sporadic ThreadSanitizer failure with OpenTracing

westonpace opened the pull request:
The particular cause of the failure was:

```
  void InputReceived(ExecNode* input, ExecBatch batch) override {
    EVENT(span_, "InputReceived", {{"batch.length", batch.length}});
    util::tracing::Span span;
    START_SPAN_WITH_PARENT(span, span_, "InputReceived",
                           {{"node.label", label()}, {"batch.length", batch.length}});

    DCHECK_EQ(input, inputs_[0]);

    bool did_push = producer_.Push(std::move(batch));
    if (!did_push) return;  // producer_ was Closed already

    if (input_counter_.Increment()) {
      // This will mark the exec plan finished
      Finish();
      // At this point the main thread is generally free to exit but span hasn't been destructed yet
    }
  }
```

It could be fixed with scoping:

```
  void InputReceived(ExecNode* input, ExecBatch batch) override {
    EVENT(span_, "InputReceived", {{"batch.length", batch.length}});
    {
      util::tracing::Span span;
      START_SPAN_WITH_PARENT(span, span_, "InputReceived",
                             {{"node.label", label()}, {"batch.length", batch.length}});

      DCHECK_EQ(input, inputs_[0]);

      bool did_push = producer_.Push(std::move(batch));
      if (!did_push) return;  // producer_ was Closed already
    }
    if (input_counter_.Increment()) {
      // This will mark the exec plan finished
      Finish();
    }
  }
```
However, I suspect this would quickly get tedious.  Instead it seems we can control the lifetime of the OT runtime storage and bind it to our worker threads.  In the future we may want to consider doing similar tricks to keep alive the memory pool, global thread pools, etc.

cyb70289 commented on cpp/src/arrow/util/tracing_internal.cc:
```diff
@@ -176,8 +176,23 @@ otel::trace::TracerProvider* GetTracerProvider() {
+
 }  // namespace
 
+static StorageSingleton storage_singleton;
```
Nit: what about moving the definition inside Attach()?

westonpace replied:
I moved the static definition inside the method (although the method names have changed a little)

cyb70289 commented on docker-compose.yml:
```diff
@@ -773,8 +773,6 @@ services:
     shm_size: *shm-size
     environment:
       <<: *ccache
-      # Bundled build of OpenTelemetry needs a git client
-      ARROW_WITH_OPENTELEMETRY: "OFF"
```
Intended or spurious change?

westonpace replied:
Good catch.  This was not intentional.

lidavidm commented on cpp/src/arrow/util/tracing_internal.cc:
```diff
@@ -176,8 +176,23 @@ otel::trace::TracerProvider* GetTracerProvider() {
+
+struct StorageSingleton {
+  StorageSingleton() : storage_(otel::context::GetDefaultStorage()) {
+    otel::context::RuntimeContext::SetRuntimeContextStorage(storage_);
```
Thanks a lot for filing this.

lidavidm replied:
Looks like the implemented it indeed. Maybe we can bump the bundled OpenTelemetry, or wait for a new release.

westonpace replied:
Looks like they are releasing monthly. Let's just disable OT in our CI builds until the release is out and we can address this at that point.

westonpace commented on cpp/src/arrow/util/thread_pool.cc:
```diff
@@ -144,6 +145,13 @@ struct ThreadPool::State {
+  // this thread ends.  This means that any calls to OpenTelemetry's static
+  // state will potentially access freed memory.  By grabbing a handle we
+  // keep OpenTelemetry's static state alive until this thread ends.
+  internal::tracing::OtHandle handle = internal::tracing::Attach();
```
I switched to what I think you were intending.  This basically inverts the dependency so that OT has to call keepalive on the thread pools instead of the other way around.  The implementation's KeepAlive method was not static though.  Can you double check that I addressed your concerns here?

westonpace commented on cpp/src/arrow/util/tracing_internal.h:
```diff
@@ -101,6 +101,17 @@ AsyncGenerator<T> WrapAsyncGenerator(AsyncGenerator<T> wrapped,
 
+class OtHandle {
+ public:
+  OtHandle(opentelemetry::nostd::shared_ptr<opentelemetry::context::RuntimeContextStorage>
```
Yes it should have been.  The implementation has changed however, all tracing_internal changes are in the anonymous namespace now.

westonpace commented on cpp/src/arrow/util/tracing_internal.h:
```diff
@@ -101,6 +101,17 @@ AsyncGenerator<T> WrapAsyncGenerator(AsyncGenerator<T> wrapped,
   };
 }
 
+class OtHandle {
```
No longer relevant.

westonpace commented:
I've made an attempt at addressing the PR feedback in preparation for the OT release.  This should be pretty close to what we will use (I think only two lines will change once we can use their new API).

The unit test is a little unreliable.  If I run it on repeat I can trigger the failure much more reliably than running it a single time.  I played around with a few different approaches but couldn't come up with a variation that failed very reliably.  I'm not sure if we want to leave it in or just get rid of it and rely on catching this via TSAN were there to be a regression.

lidavidm commented:
We can keep the test, I think. Even if we plan to catch regressions via TSAN, presumably this will trigger it more reliably?

pitrou commented:
What is the status of this?

lidavidm commented:
OpenTelemetry v1.3.0 just released with the necessary fix, so we should be able to update it and get this properly fixed now.

lidavidm commented:
@github-actions crossbow submit test-fedora-35-cpp test-ubuntu-20.04-cpp-* test-r-ubuntu-22.04

westonpace commented:
I'm pretty sure the TSAN failure is unrelated.  Just to be sure I ran this via TSAN locally and it passes fine.  I'm going to proceed with merge.
